### ISSUE TITLE
compare with saved not working

### ISSUE THREAD

**Issue by yuuuxt** [microsoft/vscode#194917]
### Applies To

- [X] Notebooks (.ipynb files)
- [ ] Interactive Window and\/or Cell Scripts (.py files with \#%% markers)

### What happened?

As shown below, the `Compare with Saved` functionality is not working for notebooks (while its' working for txt and py files).
![temp2](https://user-images.githubusercontent.com/3153534/236365145-88bf6aa1-13de-4bea-923b-92f442ff793a.gif)

### VS Code Version

1.78.0

### Jupyter Extension Version

v2023.4.1011241018

### Jupyter logs

```shell
(nothing updated there)
```

### Running Jupyter locally or remotely?

None

**Comment by DonJayamanne** [comment]
I think we should revise the solution here, and probably use the same infrastructure in `Timeline`
As the timeline has this exact same functionality (comparing unsaved/latest changes against the last saved entry).

@lszomoru What do you think about the above solution.
If not, then in the notebook land we need to have custom handling of the scheme `showModifications`, thats definitely possible, however I think thats a work around, because the `Compare With Saved` only seems to handle text files, where as `Timeline` handles custom editors natively.

**Comment by lszomoru** [comment]
> I think we should revise the solution here, and probably use the same infrastructure in Timeline
As the timeline has this exact same functionality (comparing unsaved/latest changes against the last saved entry).

@DonJayamanne, could you please elaborate? Adding @lramos15 who owns Timeline in core.

**Comment by DonJayamanne** [comment]
* the command `Compare With Saved` Registers a text content provider
* Upon selecting this command, we perform a rudimentary diff operation by comparing current content against the content on disc (the content provider will read the file off disc)
This solution is geared (hardcoded) for just plain text files.
Notebooks have a different and custom diff viewer

Time lines already support diffing against saved files (each save creates a new entry in timeline)
Hence my suggestion was to try and leverage that capability to provide this functionality.
Also timelines natively support diffing of other models (notebooks and not just plain text models).

I.e. we can compare saved file with previous changes in timeline, all we no need is the ability to compare saved file with current content.

**Comment by lszomoru** [comment]
Adding @bpasero who has implemented the "Local History" provider for timeline.

**Comment by bpasero** [comment]
What is the question here?

**Comment by DonJayamanne** [comment]
The suggestion is to re-work how `compare with saved` functionality works today. It seems to only work for text files.
Where as timelines supports other models (text, notebooks etc) and handles it in a more generic way, 
More details here https://github.com/microsoft/vscode/issues/194917#issuecomment-2461325923

**Comment by bpasero** [comment]
Yes, I think the "Compare with saved" implementation is not fit for custom editors and probably predates that work. Using a text document content provider is not good, rather it should be a custom file system so that it can work well with custom editors. 

Can you try changing it to a file system provider to see if that helps?
